lib/elements/CheckedSelect.js

```diff
@@ -71,27 +71,35 @@ class CheckedSelect extends React.Component {
             );
         });
 
+        this._visibleOptionValues = this._visibleOptions.reduce((map, option)=>{
+            map[option[this.props.valueKey]] = true;
+            return map;
+        }, {});
+
         return this._visibleOptions;
     }
 
     addVisibleOptions() {
         const values = this.props.value;
-        const valueStrings = values.map(valueObject => valueObject[this.props.valueKey]);
+        const valueStrings = values.reduce((map, valueObject)=>{
+            map[valueObject[this.props.valueKey]] = true;
+            return map;
+        }, {});
+
         // Add currently visible, enabled options to
         // the already-selected ones
         const optionsToAdd = this._visibleOptions
                 .filter(optionObject => (
                     !optionObject.disabled
-                    && valueStrings.indexOf(optionObject[this.props.valueKey]) === -1))
+                    && !valueStrings[optionObject[this.props.valueKey]]));
         return values.concat(optionsToAdd)
     }
 
     clearVisibleOptions() {
-        const visibleOptionValues = this._visibleOptions
-            .map(optionObject => optionObject[this.props.valueKey]);
+        const visibleOptionValues = this._visibleOptionValues;
         const values = this.props.value;
         // remove all visible values, keeping only invisible ones
-        return values.filter(valueObject => visibleOptionValues.indexOf(valueObject[this.props.valueKey]) === -1);
+        return values.filter(valueObject => !visibleOptionValues[valueObject[this.props.valueKey]] === -1);
```

Comment:
Should this be `!visibleOptionValues[valueObject[this.props.valueKey]]` if you don't want to use indexOf

Reply:
isn't that exactly what it says here? Not sure what you're saying

Reply:
`return values.filter(valueObject => !visibleOptionValues[valueObject[this.props.valueKey]] === -1);`

remove `=== -1`

Reply:
ah good catch